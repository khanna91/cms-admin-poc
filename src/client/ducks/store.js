import { routerMiddleware, routerReducer } from 'react-router-redux';
import { applyMiddleware, createStore, combineReducers } from 'redux';
import { createEpicMiddleware } from 'redux-observable/lib/cjs/createEpicMiddleware';
import { composeWithDevTools } from 'redux-devtools-extension';
import createHistory from 'history/createBrowserHistory';
import { combineEpics } from 'redux-observable/lib/cjs/combineEpics';


// rxjs observables
import { ajax } from 'rxjs/observable/dom/ajax';
import { of } from 'rxjs/observable/of';
// ----------------

import hocReducer from '@hox/cleanOnUnmount/reducer';
import * as reducers from '@ducks/reducers';
import * as epics from '@ducks/epics';

// global definitions
import 'rxjs/add/operator/map';
import 'rxjs/add/operator/mergeMap';
import 'rxjs/add/operator/catch';
import 'rxjs/add/operator/takeUntil';
// ------------------

export const history = createHistory();

// the epics middleware
const epicMiddleware = createEpicMiddleware({
  dependencies: {
    getJSON: ajax.getJSON,
    of
  }
});

const preloadedState = window.__PRELOADED_STATE__;
delete window.__PRELOADED_STATE__;

// the reducers
const combinedReducers = combineReducers({
  ...reducers,
  routing: routerReducer
});
const reducer = hocReducer(combinedReducers); // adds reset reducer to our apps reducer
const store = createStore(
  reducer,
  preloadedState,
  composeWithDevTools(
    applyMiddleware(
      epicMiddleware,
      routerMiddleware(history)
    )
  )
);

const allEpics = Object.values({ ...epics });
epicMiddleware.run(combineEpics(...allEpics));

export default store;
