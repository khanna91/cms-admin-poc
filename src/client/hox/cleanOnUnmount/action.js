import * as types from '@ducks/actionTypes';

/*
 * hox/cleanOnUnmount/action.js : RESET action factory
 * @returns {Object} the reset action object
 */
export function reset() { // eslint-disable-line
  return {
    type: types.RESET,
  };
}
