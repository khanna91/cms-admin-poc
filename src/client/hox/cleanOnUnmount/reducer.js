import * as types from '@ducks/actionTypes';

/*
 * hox/cleanOnUnmount/reducer.js : the cleanOnUnmount HOC reducer factory
 * @param {Function} combinedReducers the current reducers factory of the component
 * @returns {Function} combination of the current reducer factory and the new (with reset functionality) reducer factory
 */
export default combinedReducers => (state, action) => {
  if (action.type === types.RESET) {
    state = undefined; // eslint-disable-line
  }
  return combinedReducers(state, action);
};
