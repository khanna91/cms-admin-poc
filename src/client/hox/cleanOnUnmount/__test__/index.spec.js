import React from 'react';
import { render } from 'react-testing-library';
import { ConnectedRouter } from 'react-router-redux';
import { Provider } from 'react-redux';
import store, { history } from '@ducks/store';
import cleanOnUnmount from '../index';

class Test extends React.Component { // eslint-disable-line
  render() {
    return <div className="clean-on-unmount">cleanOnUnmount</div>;
  }
}

const AttachedTest = cleanOnUnmount(Test);

describe('cleanOnUnmount HOC - HOC', () => {
  it('renders with no problem', () => {
    const { container } = render(
      <Provider key="provider" store={store}>
        <ConnectedRouter key="connectedRouter" store={store} history={history}>
          <AttachedTest key="attachedTest" />
        </ConnectedRouter>
      </Provider>,
    );
    const tree = container.querySelector('.clean-on-unmount');
    expect(tree).toBeDefined();
    expect(container).toMatchSnapshot();
  });
});
