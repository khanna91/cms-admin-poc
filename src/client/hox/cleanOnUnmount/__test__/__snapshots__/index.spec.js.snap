// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`cleanOnUnmount HOC - HOC renders with no problem 1`] = `
<div>
  <div
    class="clean-on-unmount"
  >
    cleanOnUnmount
  </div>
</div>
`;
